### ISSUE TITLE
Feedback: 'Propagating Tenant Information to Downstream Endpoints'

### ISSUE THREAD

**Issue by robbaman** [Particular/docs.particular.net#4505]
Feedback for 'Propagating Tenant Information to Downstream Endpoints' https://docs.particular.net/samples/multi-tenant/propagation/

Location in GitHub: https://github.com/Particular/docs.particular.net/blob/master/samples/multi-tenant/propagation/sample.md

In the documentation you use two Behaviour classes to propagate the TenantId from incoming to outgoing message. We've previously created similar functionality ourselves using just a single mutator described below.

Seeing you as creator of the product take a different approach to this solution makes us wonder if what we're doing has some hidden downsides. Is there a reason why you split this functionality over two classes where using just the single mutator seems to work perfectly (at least functionally)

```
public class PassAlongJClubHeadersToOutgoingMessagesMutator :
	IMutateOutgoingMessages {
	public Task MutateOutgoing(MutateOutgoingMessageContext context) {
		// the outgoing headers
		var outgoingHeaders = context.OutgoingHeaders;
		if (outgoingHeaders == null) return Task.CompletedTask;
		if (!context.TryGetIncomingHeaders(out var incomingHeaders)) return Task.CompletedTask;
		
		foreach (var keyValuePair in incomingHeaders) {
			if (keyValuePair.Key.StartsWith("JC-") && !outgoingHeaders.ContainsKey(keyValuePair.Key))
				outgoingHeaders.Add(keyValuePair.Key, keyValuePair.Value);
		}

		return Task.CompletedTask;
	}
}
```

**Comment by DavidBoike** [comment]
Hi @robbaman!

Yep that will work too! Our sample uses 2 behaviors to hint how to exchange data between them, rather than always going back to the incoming message's headers as a source of truth. This enables other scenarios such as [Injecting tenant-aware components into message handlers](https://docs.particular.net/samples/multi-tenant/di/) (although I see we're going back to the message headers there too), however, it's also entirely possible that there will be multiple headers at play in a system, depending on where a message comes from, that all may contribute to a singular TenantId, which would be a good reason to have a separate incoming behavior.

So basically, there are probably a dozen different ways to attack this particular problem, and there's really nothing wrong with your implementation at all. My only nitpick is that we do generally prefer behaviors to mutators nowadays, as you get much richer access to the context. Behaviors are also compiled down at startup to be very efficient, which Daniel describes in https://particular.net/blog/10x-faster-execution-with-compiled-expression-trees

Thanks for the question!
